lib/spack/spack/test/concretize.py

```diff
@@ -437,11 +437,11 @@ def test_compiler_flags_differ_identical_compilers(self, mutable_config, clang12
                 ["hypre cflags='-g'", "^openblas"],
                 ["^openblas cflags='-g'"],
             ),
-            # Setting a flag overrides propagation
+            # TODO: should setting a flag override propagation?
             (
                 "hypre cflags=='-g' ^openblas cflags='-O3'",
                 ["hypre cflags='-g'", "^openblas cflags='-O3'"],
-                ["^openblas cflags='-g'"],
+                [],
```

Comment:
There's a minor tension here:

1. I see the appeal of reusing the `*flags=...` syntax to override propagation
  * OTOH how do we, for `x->y`, propagate flags from x to y while also adding more flags for y? I'm inclined to say we need more syntax to address this
2. Currently, the `.lp` constructs are not sufficiently fine-grained to distinguish between this and other cases checked by the `test_propagate_and_...` tests added by this PR
  * e.g. for this test to pass, `depends_on(foo cflags=...)` would disable propagation

I'm guessing the desired way to address this is to figure out how to keep track of this in `concretize.lp`: essentially node flag settings from input specs should disable propagated flags (but when flags are set by other means, propagation should be allowed). This would allow both tests to pass. I wanted to make sure this was the desired approach in case enforcing this was an afterthought, and we decide we want to support this behavior a different way.

Other ways to resolve this:

* Make `test_propagate_and_*` xfail for now (effectively, flags set via any other mechanism will disable propagation)
* Leave this (e.g. if we weren't advertising it or depending on it anywhere, don't give users a mechanism to override propagation for now)

Reply:
@becker33 after speaking with @alalazo, he mentioned these new enforcements came from a discussion with you about the intended behavior of 

```
X cflags==-g
|
Y cflags=-O3   (this is supposed to override -g, not add to it)
|
Z
```

are you particularly attached to this behavior? The current syntax must rule out either override or append; I would have assumed appending would be a more common desire.

Reply:
This is restored in [f7774eb](https://github.com/spack/spack/pull/41049/commits/f7774ebe71409ec474be0994dacf4728f22d3398)